### PR TITLE
Reworks CPR, allowing it to extend defibrillation time

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
Reworks CPR significantly. Instead of just something you can use to keep someone out of crit, it's now useful for stabilizing dead folks as well. When you successfully perform CPR on a dead person, you'll now increase their defib time by a bit. If you perform CPR soon after they died, you might get a bit of surplus time, otherwise you'll just be extending their timer for as long as you want to keep performing CPR on them. 

Without the added benefits (read: after about two minutes), performing CPR on someone while neither you nor your patient have a face covering will add four seconds (or the time it takes to perform one bout of CPR). This means that as long as someone is continually performing CPR like this, you can keep someone's defib timer paused indefinitely.
On the other hand, if either you or your patient have masks on, you'll only get two seconds of benefit for every two seconds of CPR, which means you'll be able to keep them revivable for a bit longer, but it's not going to be infinite.

As for tweaks relating to this:
- You can no longer shake dead people, only provide them with CPR (which actually brings death in line with fakedeath here).
- CPR now requires two empty hands.
- CPR can set off implanted bombs. Careful!
- CPR will now auto-repeat as long as you don't move from beside the person.

## Why It's Good For The Game
This is something that was talked about a few months ago, and it's something that I think would add a lot of value to an oft-neglected mechanic. The defib timer is usually medbay's worst enemy, and while it does lead to a lot of emergent gameplay, it can also lead to someone needing to get thrown in the cloner because they were discovered a bit too late. In this case, if someone stumbles across a fresh body, they can perform CPR until a paramedic arrives to provide actual treatment, while keeping the body from ticking down on the revive counter too much. It gives normal people a very valuable tool for helping out medbay, and a really strong reason to think about using CPR.

## Testing
Pushed some vulps under emagged recyclers until they died, watched how their revivability lasted a bit longer.

## Changelog
:cl:
tweak: CPR can now be used to extend dead people's defib timers.
tweak: CPR now requires two empty hands.
tweak: Health scanners now report the amount of time before a person will become unrevivable
tweak: CPR now auto-repeats
tweak: CPR can now set off chest-implanted bombs.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 1]
```diff
@@ -1677,49 +1677,129 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
...
 
+/// How long the mob has before they cannot be defibrillated back to life.
+/mob/living/carbon/human/defib_time_left()
+	if(stat != DEAD)
+		return 0
+	return ..() + revive_timer_postponement
+
+
+/mob/living/carbon/human/under_defib_timer()
+	return defib_time_left() > 0
```
Comment (CinnamonSnowball) [inline code/modules/mob/living/carbon/human/human_mob.dm:1688]: This override shouldn't be necessary, is it? The parent proc aready does the same exact check unless I'm missing something

**Comment by Miraviel** [comment]
question: If I understand correctly from the code, with this, you can still just remove plasmamen/vox masks and perform the full 4-seconds CPR on them without an issue, right? When they are dead, they don't breathe, so their masks being on or off doesn't really matter until they get defibbed.

unnecessary comment: the IPC exception made me laugh

**Review by CinnamonSnowball — COMMENTED** [review]
Shouldn't same-species rescue breaths work for these, at least?
[Edit] Thanks github for eating the line I meant to attack this to; meant the forced chest compressions only if either party is a vox/plasmaman change

File: code/game/objects/items/devices/scanners.dm [hunk 2]
```diff
@@ -153,8 +153,11 @@ REAGENT SCANNER
 	if(H.timeofdeath && (H.stat == DEAD || (HAS_TRAIT(H, TRAIT_FAKEDEATH))))
 		to_chat(user, "<span class='notice'>Time of Death: [station_time_timestamp("hh:mm:ss", H.timeofdeath)]</span>")
 		var/tdelta = round(world.time - H.timeofdeath)
-		if(tdelta < DEFIB_TIME_LIMIT && !DNR)
+		if(H.under_defib_timer() && !DNR)
 			to_chat(user, "<span class='danger'>Subject died [DisplayTimeText(tdelta)] ago, defibrillation may be possible!</span>")
+			if(advanced)
+				var/time_left = round(H.defib_time_left() / 10)
+				to_chat(user, "<span class='red'>Estimated time before unrecoverable tissue damage: [DisplayTimeText(time_left)].</span>")
```
Comment (hal9000PR) [inline code/game/objects/items/devices/scanners.dm:160]: maths teacher be like:
```suggestion
				to_chat(user, "<span class='red'>Estimated time before unrecoverable tissue damage: [DisplayTimeText(time_left)] seconds.</span>")
```

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/scanners.dm (hunk 2 above)
Comment (lewcc) [reply code/game/objects/items/devices/scanners.dm:160]: DisplayTimeText should display it in minutes and seconds already.

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 3]
```diff
@@ -80,3 +80,9 @@
 
 	var/list/splinted_limbs = list() //limbs we know are splinted
 	var/original_eye_color = "#000000"
+
+	/// How much extra time (beyond the defib timer) to give someone before they can't be defibrillated
+	var/revive_timer_postponement = 0
+
+	/// Whether the time of reckoning has passed and the defib timer has expired on this mob while they've been dead.
+	var/defib_timer_has_expired = FALSE
```
[the remark this answers is not in the document]
Comment (lewcc) [reply code/modules/mob/living/carbon/human/human_defines.dm:88]: Donezo. I ended up reworking revivability et al into a status effect, which you might want to take a look at.

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 4]
```diff
@@ -1686,49 +1686,152 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
...
 
 /mob/living/carbon/human/proc/do_cpr(mob/living/carbon/human/H)
+
+	var/static/list/effective_cpr_messages = list(
+		"You feel like you're able to stave off the inevitable for a little longer.",
+		"You can still see the color in their cheeks."
+	)
+
+	var/static/list/ineffective_cpr_messages = list(
+		"You're starting to feel them stiffen under you, but you keep going",
```
Comment (Henri215) [inline code/modules/mob/living/carbon/human/human_mob.dm:1703]:
```suggestion
		"You're starting to feel them stiffen under you, but you keep going.",
```

File: code/modules/mob/living/living_status_procs.dm [hunk 5]
```diff
@@ -845,6 +845,12 @@ STATUS EFFECTS
 		singlemutcheck(src, block, MUTCHK_FORCED)
 		dna.UpdateSE()
 
+// Revivability
+
+
+/mob/living/proc/IsRevivable()
```
Comment (Contrabang) [inline code/modules/mob/living/living_status_procs.dm:851]:
```suggestion
/mob/living/proc/is_revivable()
```

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 6]
```diff
@@ -1686,49 +1686,152 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
...
 
 /mob/living/carbon/human/proc/get_cpr_mod(mob/living/carbon/human/H)
 	if(is_mouth_covered() || H.is_mouth_covered())
 		return CPR_CHEST_COMPRESSION_ONLY
 	if(!H.check_has_mouth() || !check_has_mouth())
 		return CPR_CHEST_COMPRESSION_ONLY
+	if(HAS_TRAIT(src, TRAIT_NOBREATH) || HAS_TRAIT(H, TRAIT_NOBREATH))
+		return CPR_CHEST_COMPRESSION_ONLY
+
+	if(isvox(src) && !isvox(H) || !isvox(src) && isvox(H) || isplasmaman(src) && !isplasmaman(H) || !isplasmaman(src) && isplasmaman(H))  // sorry non oxy-breathers
```
Comment (S34NW) [inline code/modules/mob/living/carbon/human/human_mob.dm:1803]:
```suggestion
	if(dna.species.breathid != H.dna.species.breathid)  // sorry non oxy-breathers
```

le simplify

Comment (S34NW) [reply code/modules/mob/living/carbon/human/human_mob.dm:1803]: (please test this)

File: code/modules/mob/living/living_status_procs.dm (hunk 5 above)
Comment (lewcc) [reply code/modules/mob/living/living_status_procs.dm:851]: I was mostly following the convention of the existing status effect procs

**Review by hal9000PR — DISMISSED** [review]
would you get to seans review so we can get this merged

File: code/modules/mob/living/living_status_procs.dm (hunk 5 above)
Comment (SteelSlayer) [reply code/modules/mob/living/living_status_procs.dm:851]: Let's break that convention and go with snake case 😄

File: code/modules/mob/living/carbon/human/human_mob.dm (hunk 6 above)
Comment (lewcc) [reply code/modules/mob/living/carbon/human/human_mob.dm:1803]: I was looking all over for a variable that'd do this, thank you

Comment (lewcc) [reply code/modules/mob/living/carbon/human/human_mob.dm:1803]: Tested, it works.

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**